### PR TITLE
Save the state of the accordion using session storage

### PR DESCRIPTION
**Pull request by gemmaleigh**
When a section is opened - save its ID (with a prefix that identifies it as coming from the GOVUK Service Manual) to local storage with a value of 'Opened' and when it is closed, remove this key from session storage.

When the page loads, check session storage to see if any items have previously been stored and if so, open that section and set the `aria-expanded` attribute of the button to `true`.

cc @thehenster.

### CODE REVIEW

**Comment by gemmaleigh** [comment]
This needs tests, I'll add those.

File: app/assets/javascripts/modules/accordion-with-descriptions.js [hunk 1]
```diff
@@ -122,6 +126,53 @@
         $node.attr("aria-expanded", state);
       }
 
+      function setSessionStorage() {
+        var isOpenSubsections = $('.subsection--is-open').length;
+        if (isOpenSubsections) {
+          var $openSubsections = $('.subsection--is-open');
+          $openSubsections.each(function(index) {
+            var subsectionOpenContentId = $(this).find('.subsection__content').attr('id');
+            sessionStorage.setItem( subsectionOpenContentId , subsectionOpenContentId);
```
Comment (thehenster) [inline app/assets/javascripts/modules/accordion-with-descriptions.js]: It looks like an example is a key of `subsection_content_1` with a value of `subsection_content_1`. And this session storage is shared across the www.gov.uk subdomain. Maybe a key of `service_manual_expand_topic_subsection_1` with a value of `true` will namespace it a little and make it a bit more explanatory if someone finds it and doesn't know what it is?

File: app/assets/javascripts/modules/accordion-with-descriptions.js [hunk 2]
```diff
@@ -122,6 +126,53 @@
...
+          });
+        }
+      }
+
+      function removeSessionStorage() {
```
Comment (thehenster) [inline app/assets/javascripts/modules/accordion-with-descriptions.js]: I think I'd find this `removeSessionStorage()` and the above `setSessionStorage()` function names more representative of the role they have if they were called `setOpenTopicSections()` and `removeClosedTopicSections()`? Or `registerOpenTopicSections()` and `unregisterClosedTopicSections()`? Not sure.. Something like that maybe.

Comment (thehenster) [inline app/assets/javascripts/modules/accordion-with-descriptions.js]: Actually - ignore me. This is in the context of the accordion isn't it. I think it makes sense just like you have it.

File: app/assets/javascripts/modules/accordion-with-descriptions.js [hunk 3]
```diff
@@ -122,6 +126,53 @@
...
+            sessionStorage.removeItem( subsectionClosedContentId , subsectionClosedContentId);
+          });
+        }
+      }
+
+      function checkSessionStorage() {
+
+        // Get section IDs from Session Storage
+        var sessionStorageItems = [];
+        for ( var i = 0, len = sessionStorage.length; i < len; ++i ) {
```
Comment (thehenster) [inline app/assets/javascripts/modules/accordion-with-descriptions.js]: Session storage is shared across subdomains so if any keys are set elsewhere on GOV.UK then they'll get returned here. Maybe we could iterate over the elements in the DOM and check if they exist in storage instead?

File: app/assets/javascripts/modules/accordion-with-descriptions.js [hunk 4]
```diff
@@ -34,6 +34,17 @@
...
+      serviceManualTopic = serviceManualTopic.replace(/\s+/g,"_");
+      serviceManualTopic = serviceManualTopic.toLowerCase();
+      var GOVUKServiceManualTopic = serviceManualPrefix+"_"+serviceManualTopic;
+
+      // Add a data attribute to the wrapper with the current page's topic
+      var $subsectionWrapper = $element.find('.subsection-wrapper');
+      $subsectionWrapper.attr('data-service-manual-topic', GOVUKServiceManualTopic );
+
       // Get the open/close all button
       var $openOrCloseAllButton = $element.find('.js-subsection-controls button');
```
Comment (joelanman) [inline app/assets/javascripts/modules/accordion-with-descriptions.js]: Would allButtons plural be more readable/predictable?

Comment (joelanman) [inline app/assets/javascripts/modules/accordion-with-descriptions.js]: Sorry just realised that's not part of this pr

File: app/assets/javascripts/modules/accordion-with-descriptions.js [hunk 5]
```diff
@@ -122,6 +144,59 @@
...
+          sessionStorageItems.push(sessionStorage.key( i ));
+        }
+
+        if (sessionStorageItems.length > 0) {
+          $.each( sessionStorageItems, function( key, value ) {
+            // console.log("key:"+key, "value:"+value);
+            // If any items in session storage start with GOVUK_Service_Manual and the topic name
+            if (value.indexOf(GOVUKServiceManualTopic) >= 0){
+              // console.log("items matched with prefix:"+GOVUKServiceManualTopic);
+              var subsectionContentId = value.toString();
```
Comment (joelanman) [inline app/assets/javascripts/modules/accordion-with-descriptions.js]: Is this needed? I think local storage can only store strings

File: app/assets/javascripts/modules/accordion-with-descriptions.js [hunk 6]
```diff
@@ -34,6 +34,17 @@
         ariaControlsValue += "subsection_content_"+i+" ";
       }
 
+      // Create a unique prefix for each service manual topic
+      var serviceManualPrefix = "GOVUK_service_manual";
+      var serviceManualTopic = $('h1').text();
+      serviceManualTopic = serviceManualTopic.replace(/\s+/g,"_");
+      serviceManualTopic = serviceManualTopic.toLowerCase();
+      var GOVUKServiceManualTopic = serviceManualPrefix+"_"+serviceManualTopic;
```
Comment (joelanman) [inline app/assets/javascripts/modules/accordion-with-descriptions.js]: why add GOVUK prefix to the variable name at this point?

Comment (gemmaleigh) [inline app/assets/javascripts/modules/accordion-with-descriptions.js]: This variable is going to be saved in session storage, so it has been added to make it clear where these values have come from.

File: app/assets/javascripts/modules/accordion-with-descriptions.js (hunk 4 above)
Comment (gemmaleigh) [inline app/assets/javascripts/modules/accordion-with-descriptions.js]: It's only one button :)

Comment (joelanman) [inline app/assets/javascripts/modules/accordion-with-descriptions.js]: ha ha! I read it wrong :)

File: app/assets/javascripts/modules/accordion-with-descriptions.js [hunk 7]
```diff
@@ -8,51 +8,81 @@
...
+      initialize();
 
-      var $subsectionButton = $element.find('.subsection__button');
-      var $subsectionHeader = $element.find('.subsection__header');
+      function initialize() {
+        addOpenCloseAllButton();
+        addSubsectionButtons();
+        addSubsectionIcons();
+        updateSubsectionControls();
+        getServiceManualTopicPrefix();
```
Comment (thehenster) [inline app/assets/javascripts/modules/accordion-with-descriptions.js]: `getServiceManualTopicPrefix()` just returns a value so there shouldn't be a need to call it here.

Comment (thehenster) [inline app/assets/javascripts/modules/accordion-with-descriptions.js]: Ah - `GOVUKServiceManualTopic` is defined outside of the function `getServiceManualTopicPrefix()`. So there isn't any need for the return statement in `getServiceManualTopicPrefix()`.
